include/rdma/fi_endpoint.h

```diff
@@ -66,6 +66,7 @@ enum {
 	FI_OPT_RECV_BUF_SIZE,
 	FI_OPT_TX_SIZE,
 	FI_OPT_RX_SIZE,
+	FI_OPT_RNR_RETRY,
```

Comment:
RNR is not an API related definition.  This is a provider internal implementation detail.

Reply:
I will move this to efa provider specific definition.

Reply:
@shefty, if customer expects to reset RNR retry counter via fi_setopt, can we define a range for providers to define their specific options?

Reply:
Yes, you should be able to use FI_PROV_SPECIFIC to set the high-order bit indicating a provider specific value.

Reply:
@shefty I'm not sure I fully understand your suggestion, so could be easily wrong here. 
Is it something like this? 
```
--- a/include/rdma/fi_endpoint.h
+++ b/include/rdma/fi_endpoint.h
@@ -66,6 +66,7 @@ enum {
        FI_OPT_RECV_BUF_SIZE,
        FI_OPT_TX_SIZE,
        FI_OPT_RX_SIZE,
+       FI_OPT_EFA_RNR_RETRY = (1U | FI_PROV_SPECIFIC),
 };
```
Or we need to create a provider-specific header file, define FI_OPT_EFA_RNR_RETRY in it, and apps need to include this new header file to properly call fi_setopt with the defined optname? Basically follow the similar approach of https://github.com/ofiwg/libfabric/blob/main/prov/usnic/src/fi_ext_usnic.h

Reply:
Let me respond on the mailing list

Reply:
Thanks, @shefty. I think we reach to a conclusion on the mailing list. In my case, I will
1. In fi_endpoint.h, define a FI_OPT_PROV_SPECIFIC  (1U << 30)
2. In fi_ext.h, define FI_OPT_EFA_RNR_RETRY (1U | FI_OPT_PROV_SPECIFIC)
Am i understanding correctly? 
If you have a generic example showing as a guideline, pls let me know, I can add it as the comment for other providers to follow.

Reply:
I would use the existing FI_PROV_SPECIFIC definition, versus also defining FI_OPT_PROV_SPECIFIC.

I was also considering a provider specific code in fi_ext.h that could ensure 2 providers don't select the same values.  E.g.
```
#define FI_PROV_SPECIFIC_EFA (FI_PROV_SPECIFIC | (0xefa << 20))
#define FI_PROV_SPECIFIC_VERBS (FI_PROV_SPECIFIC | 0xd3a << 20)

/* EFA options */
enum {
    FI_OPT_EFA_RNR_RETRY (1U | FI_PROV_SPECIFIC_EFA)
};
```
Not sure if the provider codes should be 8, 12, or 16 bits, or if it really matters much.

Reply:
One thing I think we should consider is that using FI_PROV_SPECIFIC (#define FI_PROV_SPECIFIC (1U << 31)) could exceed the range `int` allows, as optname is type of integer in definition (`fi_setopt(fid_t fid, int level, int optname, const void *optval, size_t optlen)`)

Reply:
Are we really dealing with any processor that doesn't have a 32-bit integer?

Reply:
No, that's just a minor thought. Let's ignore it. 
For provider codes, 12 or 16 bits looks good to me,  which allows a "meaningful" provider code and also leave large enough space (20 or 16 bits) for provider specific values.

Reply:
changes are in #7065